tests/AppBundle/Listener/ListenerTest.php

```diff
@@ -0,0 +1,152 @@
+<?php
+
+namespace Tests\AppBundle\Listener;
+
+use AppBundle\Listener\AccessListener;
+use AppBundle\Listener\ResponseListener;
+use AppBundle\Security\RateLimiter;
+use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Controller\ArgumentResolver;
+use Symfony\Component\HttpKernel\Controller\ControllerResolver;
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\HttpKernel;
+
+class ListenerTest extends KernelTestCase
+{
+    /**
+     * @var \Doctrine\ORM\EntityManager
+     */
+    private $em;
+
+    protected function setUp()
+    {
+        self::bootKernel();
+
+        $this->em = static::$kernel->getContainer()
+            ->get('doctrine')
+            ->getManager();
+    }
+
+    private function getResponseEvent()
+    {
+        $request = Request::create(
+            '/videos',
+            'GET'
+        );
+        $dispatcher = new EventDispatcher();
+
+        $controllerResolver = new ControllerResolver();
+        $argumentResolver = new ArgumentResolver();
+
+        $kernel = new HttpKernel($dispatcher, $controllerResolver, new RequestStack(), $argumentResolver);
+
+        $event = new GetResponseEvent($kernel, $request, 1);
```

Comment:
Can't you just mock the `GetResponseEvent` instance and return the `$request` when the `getRequest()` method is called and `true` when the `isMasterRequest()` is called.

You will just have to check if `setResponse` has been called on the mock or not:

```php
$event->expects($this->once()) // or $this->none()
                 ->method('setResponse')
```

If setResponse has been called, it is a too many request. If it has not been called, the request you've created is considered as valid.